src/main/java/org/mockito/internal/util/MockUtil.java

```diff
@@ -18,22 +19,68 @@
 import org.mockito.plugins.MockMaker.TypeMockability;
 import org.mockito.plugins.MockResolver;
 
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
 
 import static org.mockito.internal.handler.MockHandlerFactory.createMockHandler;
 
 @SuppressWarnings("unchecked")
 public class MockUtil {
 
-    private static final MockMaker mockMaker = Plugins.getMockMaker();
+    private static final MockMaker defaultMockMaker = Plugins.getMockMaker();
+    private static final Map<Class<? extends MockMaker>, MockMaker> mockMakers =
+            new ConcurrentHashMap<>(
+                    Collections.singletonMap(defaultMockMaker.getClass(), defaultMockMaker));
+    private static final DefaultMockitoPlugins defaultMockitoPlugins = new DefaultMockitoPlugins();
 
     private MockUtil() {}
 
-    public static TypeMockability typeMockabilityOf(Class<?> type) {
-        return mockMaker.isTypeMockable(type);
+    private static MockMaker getMockMaker(String mockMaker) {
+        if (mockMaker == null) {
+            return defaultMockMaker;
+        }
+
+        String typeName;
+        if (DefaultMockitoPlugins.MOCK_MAKER_ALIASES.contains(mockMaker)) {
+            typeName = defaultMockitoPlugins.getDefaultPluginClass(mockMaker);
+        } else {
+            typeName = mockMaker;
+        }
+
+        Class<? extends MockMaker> type;
+        // TODO: #2701 I doubt that it is a good idea to use the context class loader here,
+        //  but PluginInitializer.loadImpl is using it as well,
+        //  but DefaultMockitoPlugins.create is explicitly not using it.
+        ClassLoader loader = Thread.currentThread().getContextClassLoader();
```

Comment:
Using `Thread.currentThread().getContextClassLoader()` could cause problems with our built-in mock makers (see pre-existing comment in [`DefaultMockitoPlugins.create`](https://github.com/mockito/mockito/blob/530558ae3c721cb310f3410a66ff4061414cf836/src/main/java/org/mockito/internal/configuration/plugins/DefaultMockitoPlugins.java#L80-L83)). It might also cause a memory leak if someone is loading a lot of mock makers from temporary class loaders.

However, `PluginInitializer.loadImpl` is also using the context class loader, which is even more problematic in my opinion. The result of `PluginInitializer.loadImpl` is stored in static fields (i.e. `MockUtil.defaultMockMaker` and `Plugins.registry.mockMaker`), so Mockito considers the context class loader only when these static fields are first initialized. This probably means that the test which is executed first determines the mock maker which is used as the default, which seems quite arbitrary to me.

@TimvdLippe, do you know if there is any history about the usage of context class loaders with Mockito? Or do you have any opinion on that?

Reply:
No I am not ware of any history. Usually, if there are issues with ClassLoaders, we write specific integration tests and then we fix them. In this case, since I am not aware of any problems, I think it's fine to keep it as-is.